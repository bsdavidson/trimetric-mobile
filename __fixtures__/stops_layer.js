export const stopPoints = {
  type: "FeatureCollection",
  features: [
    {
      type: "Feature",
      properties: {type: "stop", stop_id: "13990"},
      geometry: {type: "Point", coordinates: [-122.398702, 45.552013]}
    },
    {
      type: "Feature",
      properties: {type: "stop", stop_id: "13991"},
      geometry: {type: "Point", coordinates: [-122.407683, 45.55363]}
    }
  ]
};

export const stops = [
  {
    id: "2",
    code: "2",
    name: "A Ave & Chandler",
    desc: "Eastbound stop in Lake Oswego (Stop ID 2)",
    lat: 45.420609,
    lng: -122.675671,
    zone_id: "B",
    url: "http://trimet.org/#tracker/stop/2",
    location_type: 0,
    parent_station: "",
    direction: "East",
    position: "Nearside",
    wheelchair_boarding: 0,
    distance: 0
  },
  {
    id: "landmark-2560",
    code: "",
    name: "SE Park Ave MAX Station",
    desc: "",
    lat: 45.430735,
    lng: -122.635065,
    zone_id: "",
    url: "",
    location_type: 1,
    parent_station: "",
    direction: "",
    position: "",
    wheelchair_boarding: 0,
    distance: 0
  }
];
