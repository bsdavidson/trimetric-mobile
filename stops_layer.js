import React, {Component} from "react";
import {Platform, StyleSheet, Text, View} from "react-native";
import Mapbox from "@mapbox/react-native-mapbox-gl";

import {MIN_LABEL_LAYER_ID} from "./App";

export function StopsLayer(stopPoints, filter) {
  if (!stopPoints) {
    return null;
  }
  return (
    <View>
      <Mapbox.ShapeSource
        id="stop_symbols_source"
        shape={stopPoints}
        images={{
          assets: ["stop"]
        }}>
        <Mapbox.SymbolLayer
          id="stop_symbols_layer"
          minZoomLevel={13}
          filter={
            filter ? ["==", filter.type, filter.value] : ["!=", "test", "123"]
          }
          belowLayerID={MIN_LABEL_LAYER_ID}
          style={mapStyles.stopSymbolsLayer}
        />
      </Mapbox.ShapeSource>

      <Mapbox.ShapeSource id="stop_points_source" shape={stopPoints}>
        <Mapbox.CircleLayer
          id="stop_points_layer"
          maxZoomLevel={13}
          filter={
            filter ? ["==", filter.type, filter.value] : ["!=", "test", "123"]
          }
          belowLayerID={MIN_LABEL_LAYER_ID}
          style={mapStyles.stopPointsLayer}
        />
      </Mapbox.ShapeSource>
    </View>
  );
}

const mapStyles = Mapbox.StyleSheet.create({
  stopPointsLayer: {
    circleColor: "#555555",
    circleOpacity: 0.8,
    circleRadius: Mapbox.StyleSheet.camera(
      {
        0: 0,
        9: 1,
        10: 1.2,
        13: 3
      },
      Mapbox.InterpolationMode.Linear
    )
  },
  stopSymbolsLayer: {
    iconImage: "marker-15",
    iconAllowOverlap: true,

    iconSize: Mapbox.StyleSheet.camera(
      {
        13: 0.5,
        18: 1.5
      },
      Mapbox.InterpolationMode.Linear
    )
  }
});
