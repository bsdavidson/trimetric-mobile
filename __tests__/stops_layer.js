import "react-native";
import React from "react";
import renderer from "react-test-renderer";

import {mapStateToProps, StopsLayer} from "../stops_layer";
import {stopPoints, stops} from "../__fixtures__/stops_layer.js";

jest.mock("@mapbox/react-native-mapbox-gl");

it("renders StopsLayer", () => {
  const slNull = renderer.create(<StopsLayer />).toJSON();
  expect(slNull).toMatchSnapshot("null_stopsLayer");

  // NonNull
  const sl = renderer.create(<StopsLayer stopPoints={stopPoints} />).toJSON();
  expect(sl).toMatchSnapshot("non_null_stopsLayer");
});

it("has mapStateToProps", () => {
  // Filter nothing
  let state = mapStateToProps({
    layerVisibility: {
      stops: true
    },
    selectedItems: {
      features: []
    },
    stops: []
  });
  expect(state).toEqual({
    filter: ["!=", "non_existing_attribute", "1"],
    stopPoints: {features: [], type: "FeatureCollection"}
  });

  // Filter all stops
  state = mapStateToProps({
    layerVisibility: {
      stops: false
    },
    selectedItems: {
      features: []
    },
    stops: []
  });
  expect(state).toEqual({
    filter: ["==", "non_existing_attribute", "1"],
    stopPoints: {features: [], type: "FeatureCollection"}
  });

  // Filter to specific stop
  state = mapStateToProps({
    selectedArrival: {somethingNotNull: "1"},
    layerVisibility: {
      stops: true
    },
    selectedItems: {
      features: [
        {
          properties: {
            type: "stop",
            stop_id: "2" // First item in stops in fixtures
          },
          geometry: {
            coordinates: [-122.675671, 45.420609]
          }
        }
      ]
    },
    selectedItemIndex: 0,
    stops: stops
  });
  expect(state).toEqual({
    filter: ["==", "stop_id", "2"],
    stopPoints: {
      features: [
        {
          geometry: {coordinates: [-122.675671, 45.420609], type: "Point"},
          properties: {stop_id: "2", type: "stop"},
          type: "Feature"
        },
        {
          geometry: {coordinates: [-122.635065, 45.430735], type: "Point"},
          properties: {stop_id: "landmark-2560", type: "stop"},
          type: "Feature"
        }
      ],
      type: "FeatureCollection"
    }
  });
});
