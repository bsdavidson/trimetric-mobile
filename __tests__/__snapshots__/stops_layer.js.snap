// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`non_null_stopsLayer 1`] = `
<View>
  <ShapeSource
    id="stop_symbols_source"
    images={
      Object {
        "assets": Array [
          "stop",
        ],
      }
    }
    shape={
      Object {
        "features": Array [
          Object {
            "geometry": Object {
              "coordinates": Array [
                -122.398702,
                45.552013,
              ],
              "type": "Point",
            },
            "properties": Object {
              "stop_id": "13990",
              "type": "stop",
            },
            "type": "Feature",
          },
          Object {
            "geometry": Object {
              "coordinates": Array [
                -122.407683,
                45.55363,
              ],
              "type": "Point",
            },
            "properties": Object {
              "stop_id": "13991",
              "type": "stop",
            },
            "type": "Feature",
          },
        ],
        "type": "FeatureCollection",
      }
    }
  >
    <SymbolLayer
      belowLayerID="place-city-sm"
      filter={undefined}
      id="stop_symbols_layer"
      minZoomLevel={13}
      style={
        Object {
          "iconAllowOverlap": true,
          "iconImage": "marker-15",
          "iconSize": "camera_value",
        }
      }
    />
  </ShapeSource>
  <ShapeSource
    id="stop_points_source"
    shape={
      Object {
        "features": Array [
          Object {
            "geometry": Object {
              "coordinates": Array [
                -122.398702,
                45.552013,
              ],
              "type": "Point",
            },
            "properties": Object {
              "stop_id": "13990",
              "type": "stop",
            },
            "type": "Feature",
          },
          Object {
            "geometry": Object {
              "coordinates": Array [
                -122.407683,
                45.55363,
              ],
              "type": "Point",
            },
            "properties": Object {
              "stop_id": "13991",
              "type": "stop",
            },
            "type": "Feature",
          },
        ],
        "type": "FeatureCollection",
      }
    }
  >
    <CircleLayer
      belowLayerID="place-city-sm"
      filter={undefined}
      id="stop_points_layer"
      maxZoomLevel={13}
      style={
        Object {
          "circleColor": "#555555",
          "circleOpacity": 0.8,
          "circleRadius": "camera_value",
        }
      }
    />
  </ShapeSource>
</View>
`;

exports[`null_stopsLayer 1`] = `null`;
